core-processor/src/common.rs

```diff
@@ -0,0 +1,219 @@
+
+use crate::{executor::ERR_EXIT_CODE, id};
+
+use gear_core::{
+    memory::PageNumber,
+    message::{Message, MessageId},
+    program::{Program, ProgramId},
+};
+
+#[derive(Clone, Debug)]
+pub enum DispatchKind {
+    Init,
+    Handle,
+    HandleReply,
+}
+
+impl DispatchKind {
+    pub fn into_entry(self) -> &'static str {
+        match self {
+            Self::Init => "init",
+            Self::Handle => "handle",
+            Self::HandleReply => "handle_reply",
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Dispatch {
+    pub kind: DispatchKind,
+    pub message: Message,
+}
+
+#[derive(Clone)]
+pub enum DispatchResultKind {
+    Success,
+    Trap(Option<&'static str>),
+    Wait,
+}
+
+pub struct DispatchResult {
+    pub kind: DispatchResultKind,
+
+    pub program: Program,
+    pub dispatch: Dispatch,
+
+    pub messages: Vec<Message>,
+    pub awakening: Vec<MessageId>,
+
+    pub gas_left: u64,
+    pub gas_burned: u64,
+
+    pub page_update: BTreeMap<PageNumber, Vec<u8>>,
+    pub nonce: u64,
+}
+
+impl DispatchResult {
+    pub fn program(&self) -> Program {
+        self.program.clone()
+    }
+
+    pub fn message_source(&self) -> ProgramId {
+        self.dispatch.message.source()
+    }
+
+    pub fn message_nonce(&self) -> u64 {
+        self.nonce
+    }
+
+    pub fn fetch_inc_message_nonce(&mut self) -> u64 {
+        let nonce = self.nonce;
+        self.nonce += 1;
+        nonce
+    }
+
+    pub fn apply_nonce(&mut self) {
+        self.program.set_message_nonce(self.nonce)
+    }
+
+    pub fn program_id(&self) -> ProgramId {
+        self.program.id()
+    }
+
+    pub fn message_id(&self) -> MessageId {
+        self.dispatch.message.id()
+    }
+
+    pub fn dispatch(&self) -> Dispatch {
+        self.dispatch.clone()
+    }
+
+    pub fn kind(&self) -> DispatchResultKind {
+        self.kind.clone()
+    }
+
+    pub fn gas_left(&self) -> u64 {
+        self.gas_left
+    }
+
+    pub fn gas_burned(&self) -> u64 {
+        self.gas_burned
+    }
+
+    pub fn outgoing(&self) -> Vec<Message> {
+        self.messages.clone()
+    }
+
+    pub fn awakening(&self) -> Vec<MessageId> {
+        self.awakening.clone()
+    }
+
+    pub fn page_update(&self) -> BTreeMap<PageNumber, Vec<u8>> {
+        self.page_update.clone()
+    }
+
+    pub fn trap_reply(&mut self) -> Option<Message> {
+        if let Some((_, exit_code)) = self.dispatch.message.reply() {
+            if exit_code != 0 {
+                return None;
+            }
+        };
+
+        Some(Message::new_reply(
+            id::next_message_id(self.program_id(), self.fetch_inc_message_nonce()),
+            self.program_id(),
+            self.dispatch.message.source(),
+            Default::default(),
+            0,
+            0,
+            self.message_id(),
+            ERR_EXIT_CODE,
+        ))
+    }
+}
+
+#[derive(Debug)]
+pub enum JournalNote {
+    ExecutionFail {
+        origin: MessageId,
```

Comment:
What do you think about changing field name? `origin` is related to substrate's account/message origin concept, so can lead to misunderstanding.